libs/@local/hash-graph-types/graph_types/_const.py

```diff
@@ -0,0 +1,70 @@
+from typing import Annotated, Any, TypeVar, cast, overload, Literal
+
+from pydantic import GetCoreSchemaHandler, GetJsonSchemaHandler
+from pydantic.json_schema import JsonSchemaValue
+from pydantic_core import CoreSchema, core_schema
+
+T = TypeVar("T")
+
+
+class ConstAnnotation:
+    const: Any
+
+    @classmethod
+    def __get_pydantic_core_schema__(
+        cls,
+        source_type: Any,  # noqa: ANN401
+        handler: GetCoreSchemaHandler,
+    ) -> CoreSchema:
+        schema = handler(source_type)
+
+        return core_schema.no_info_after_validator_function(
+            cls.validate_const,
+            schema,
+        )
+
+    @classmethod
+    def __get_pydantic_json_schema__(
+        cls,
+        schema: CoreSchema,
+        handler: GetJsonSchemaHandler,
+    ) -> JsonSchemaValue:
+        json_schema = handler(schema)
+        json_schema.update(const=cls.const)
+
+        return json_schema
+
+    @classmethod
+    def validate_const(cls, v: T) -> T:
+        if v != cls.const:
+            msg = f"Value must be {cls.const}"
+            raise ValueError(msg)
+
+        return v
+
+
+_Undefined = object()
+
+
+@overload
+def constant(const_: T, /) -> type[T]:
```

Comment:
What does the `/` mean?

Reply:
Every argument before `/` is required to be positional

Reply:
We use `*` for this in other locations, is it the same as `*, const_: T`?

Reply:
Nope,

Arguments before `/` are required positional, arguments after `*` are required keyword. Arguments in between them are positional-or-keyword.

So it's `(positional-only *)? positional-or-keyword (* keyword-only)?`